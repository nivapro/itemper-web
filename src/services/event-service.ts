import Vue from 'vue';

export interface IEventService {
    on(event: string | string[], callback: (...args: unknown[]) => void): this;
    once(event: string | string[], callback: (...args: unknown[]) => void): this;
    off(event: string | string[], callback: (...args: unknown[]) => void): this;
    emit(event: string | string[], ...args: unknown[]): this;
}
export class EventService implements IEventService {
    private eventBus = new Vue();

    public on(event: string  | string[], callback: (...args: unknown[]) => void): this {
        this.eventBus.$on(event, callback);
        return this;
    }
    public once(event: string | string[], callback: (...args: unknown[]) => void): this {
        this.eventBus.$once(event, callback);
        return this;
    }
    public off(event: string | string[], callback: (...args: unknown[]) => void): this {
        this.eventBus.$off(event, callback);
        return this;
    }
    public emit(event: string, ...args: unknown[]): this {
        this.eventBus.$emit(event, ...args);
        return this;
    }
}

