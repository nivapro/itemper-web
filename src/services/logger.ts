import * as winston from 'winston';
const transports = {
//    file: new winston.transports.File({ filename: 'itemper-vue-error.log', level: 'error' }),
    console: new (winston.transports.Console)(),
};

export const log =  new winston.Logger({
    exitOnError: false,
    level: 'info',
    transports: [
  //      transports.file,
        transports.console,
    ],
  });

export function setLevel(level: string): void {
    log.transports.console.level = level;
}
  // let v: winston.LoggerOptions;
// if we're not in production then log to the `console` with the format:
// `${info.level}: ${info.message} JSON.stringify({ ...rest }) `
//
// if (process.env.NODE_ENV !== 'production') {
// logger.add(new winston.transports.Console({
//     format: winston.configure({

//     })
// }));
// }
export default log;
