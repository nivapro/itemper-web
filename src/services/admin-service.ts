import { LogLevel } from '@/models/admin';

import { IApiService, Method } from '@/services/api-service';

export interface IAdminService {
        setLogLevel(level: LogLevel): Promise<boolean>;
}

export class AdminService implements IAdminService {

    private api: IApiService;
    private path = '/admin';
    constructor(apiService: IApiService) {
        this.api = apiService;
    }
    public setLogLevel(newLevel: LogLevel): Promise<boolean> {
        const method: Method = 'put';
        const path = this.path;
        const body = { level: LogLevel[newLevel] };
        return this.api.request(method, path, body);
    }
}
