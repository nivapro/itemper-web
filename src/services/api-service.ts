import axios, {  Method, AxiosInstance, AxiosRequestConfig } from 'axios';

import { iTemperAPI } from '@/config';
import { json } from '@/helpers/';
import { log } from '@/services/logger';

export interface IApiService {
    isLoggedIn: boolean;
    register(email: string, password: string, confirmPassword: string): Promise<boolean>;
    login(email: string, password: string): Promise<boolean>;
    logout(): void;
    request(method: Method, url: string, body?: any, config?: AxiosRequestConfig): Promise<any>;
}

export interface IError {
    message: string;
    status: number;
}

export { Method } from 'axios';

export class ApiService implements IApiService {

    public isLoggedIn: boolean = false;

    private io: AxiosInstance;
    private token: string | undefined = undefined;
    private ContentTypeHeader = {'Content-Type': 'application/json'};

    constructor() {
        this.io = axios.create({
            baseURL: iTemperAPI,
            headers: this.ContentTypeHeader,
        });
    }

    public register(email: string, password: string, confirmPassword: string): Promise<boolean> {
        log.debug('ApiService.register');
        const url = '/signup';
        const body = JSON.stringify({email, password, confirmPassword});
        return this.post(url, body);
    }
    public login(email: string, password: string): Promise<boolean> {
        log.debug('ApiService.login');
        const url = '/login';
        const body = JSON.stringify({email, password});
        return this.post(url, body);
    }

    public logout() {
        log.debug('ApiService.logout');
        this.token = undefined;
        this.isLoggedIn = false;
    }

    public request(method: Method, url: string, body?: any, config?: AxiosRequestConfig): Promise<any> {
        log.debug('ApiService.request: ' + method.toUpperCase() + ' ' + url);
        return new Promise<any> ((resolve, reject) => {
            if (!this.isLoggedIn) {
                    reject('User is not logged');
            }
            let conf: AxiosRequestConfig;
            const Authorization = {Authorization: this.Authorization().value};
            if (config) {
                conf = config;
                conf.method = method;
                conf.url = url;
                if (conf.headers) {
                    conf.headers.Authorization = this.Authorization().value;
                } else {
                    conf.headers = Authorization;
                }
                if (body) {
                    conf.data = body;
                }
            } else {
                conf = {url, method, headers: Authorization, data: body};
            }
            this.io.request(conf)
            .then ((response) => {
                    const data = response.data;
                    resolve(data);
            })
            .catch((error: any) => {
                reject(this.createError(error));
            });
        });
    }
    private createError(error: any): IError {

        let message = 'Unknown error';
        let status = 99;
        if (error.message) {
            message =  error.message;
            status = 98;
        }
        if (error.response) {
            status  = error.response.status;
            if (status === 422) {
                message = 'Invalid field value(s)';
            } else {
                message = error.response.data;
            }
        } else if (error.request) {
            log.debug(error.request);
            message = error.request.data;
            status = error.request.status;
        } else {
            log.debug('request Error=' + error.message);
        }
        return {message, status};
    }
    private Authorization() {
        log.debug('ApiService.Authorization');
        if (this.token) {
            return {value: 'Bearer ' + this.token};
        } else {
            return { value: 'Bearer '};
        }
    }
    private post(url: string, body: string): Promise<boolean> {
        return new Promise<boolean> ((resolve, reject) => {
            this.io.post(url, body)
            .then((response) => {
                log.debug(json(response));
                this.token = response.data.token;
                this.isLoggedIn = true;
                resolve(true);
            })
            .catch((error: any) => {
                this.isLoggedIn = false;
                reject(this.createError(error));
            });
        });
    }
}
