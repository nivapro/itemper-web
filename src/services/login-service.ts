import axios, { AxiosInstance } from 'axios';

import { User } from '@/models/user';

import { iTemperAPI } from '@/config';
import { json } from '@/helpers/';
import { log } from '@/services/logger';

export interface ILoginService {
    isLoggedIn: boolean;
    register(email: string, password: string, confirmPassword: string): Promise<User>;
    login(email: string, password: string): Promise<User>;
    logout(): void;
    Authorization(): { value: string};
}

export class LoginService implements ILoginService {

    public isLoggedIn: boolean = false;

    private io: AxiosInstance;
    private token: string | undefined = undefined;
    private ContentTypeHeader = {'Content-Type': 'application/json'};

    constructor() {
        this.io = axios.create({headers: this.ContentTypeHeader});
    }

    // public get isLoggedIn(): boolean {
    //     return this.token !== undefined;
    // }
    public register(email: string, password: string, confirmPassword: string): Promise<User> {
        log.debug('LoginService.register');
        const url = iTemperAPI + '/signup';
        const body = JSON.stringify({email, password, confirmPassword});
        return this.post(url, body);
    }
    public login(email: string, password: string): Promise<User> {
        log.debug('LoginService.login');
        const url = iTemperAPI + '/login';
        const body = JSON.stringify({email, password});
        return this.post(url, body);
    }

    public logout() {
        log.debug('LoginService.logout');
        this.token = undefined;
        this.isLoggedIn = false;
    }

    public Authorization() {
        log.debug('LoginService.Authorization');
        if (this.token) {
            return {value: 'Bearer ' + this.token};
        } else {
            return { value: 'Bearer '};
        }
    }

    private post(url: string, body: string): Promise<User> {
        return new Promise<User> ((resolve, reject) => {
            this.io.post(url, body)
            .then((response) => {
                this.token = response.data.token;
                const email = response.data.email;
                const password = response.data.password;
                const user = new User(email, password);
                this.isLoggedIn = true;
                resolve(user);
            })
            .catch((err: any) => {
                log.debug('LoginService.login: error=' + json(err.response));
                reject(err.response);
            });
        });
    }
}
