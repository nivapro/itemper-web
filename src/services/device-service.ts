import { Device, DeviceData } from '@/features/devices';

import { IApiService, Method } from '@/services/api-service';

export interface IDeviceService {
        getDevices(): Promise<DeviceData[]>;
        createDevice(name: string, color: string): Promise<DeviceData>;
        renameDevice(name: string, device: Device): Promise<DeviceData>;
        updateColor(color: string, device: Device): Promise<DeviceData>;
        deleteDevice(device: Device): Promise<DeviceData>;
}

export class DeviceService implements IDeviceService {

    private api: IApiService;
    private path = '/devices';
    constructor(apiService: IApiService) {
        this.api = apiService;
    }

    public getDevices(): Promise<DeviceData[]> {
        const method: Method = 'get';
        return this.api.request(method, this.path);
    }
    public createDevice(name: string, color: string): Promise<DeviceData> {
        const method: Method = 'post';
        return this.api.request(method, this.path, {name, color});
    }
    public renameDevice(name: string, device: Device): Promise<DeviceData> {
        const path = this.path + '/' + device.deviceID + '/name';
        const method: Method = 'put';
        return this.api.request(method, path, {name});
    }
    public updateColor(color: string, device: Device): Promise<DeviceData> {
        const path = this.path + '/' + device.deviceID + '/color';
        const method: Method = 'put';
        return this.api.request(method, path, {color});
    }
    public deleteDevice(device: Device): Promise<DeviceData> {
        const path = this.path + '/' + device.deviceID;
        const method: Method = 'delete';
        return this.api.request(method, path);
    }
}
