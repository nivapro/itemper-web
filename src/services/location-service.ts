import { iTemperAPI } from '@/config';
import { Device} from '@/models/device';

import {json} from '@/helpers';
import { log } from '@/services/logger';
import { IApiService, Method } from '@/services/api-service';

export interface ILocationService {
        getDevices(): Promise<Device[]>;
        createDevice(name: string): Promise<Device>;
        renameDevice(name: string, device: Device): Promise<Device>;
        deleteDevice(device: Device): Promise<Device>;
}

export class LocationService implements ILocationService {

    private api: IApiService;

    constructor(apiService: IApiService) {
        this.api = apiService;
    }

    public getDevices(): Promise<Device[]> {
            const url = iTemperAPI + '/devices';
            const method: Method = 'get';
            return this.request(method, url);
    }
    public createDevice(name: string): Promise<Device> {
        const url = iTemperAPI + '/devices';
        const method: Method = 'post';
        return this.request(method, url, {name});
    }
    public renameDevice(name: string, device: Device): Promise<Device> {
        const url = iTemperAPI + '/devices/' + device.deviceID;
        const method: Method = 'put';
        return this.request(method, url, {name});
    }
    public deleteDevice(device: Device): Promise<Device> {
        const url = iTemperAPI + '/devices/' + device.deviceID;
        const method: Method = 'delete';
        return this.request(method, url);
    }

    private request(method: Method, url: string, body?: any): Promise<any> {
        log.debug('DeviceService.request: ' + method.toUpperCase() + ' ' + url);
        return new Promise<any> ((resolve, reject) => {
                if (!this.api.isLoggedIn) {
                        reject('DeviceService.request: user is not logged');
                }
                this.api.request(method, url, body)
                .then ((response) => {
                        const data = response.data.slice();
                        log.debug('DeviceService.getDevices:  axios - response sensors=' + json(data));
                        resolve(data);
                })
                .catch((error: any) => {
                        reject(error);
                });
        });
    }
}
