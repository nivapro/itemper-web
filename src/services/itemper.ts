import { ApiService } from './api-service';
import { AdminService } from './admin-service';
import { DeviceService } from './device-service';
import { EventService } from './event-service';
import { LocationService } from '../features/locations';
import { SensorService } from './sensor-service';

export interface IiTemper {
    apiService: ApiService;
    deviceService: DeviceService;
    locationService: LocationService;
    sensorService: SensorService;
}
export class Itemper implements IiTemper {
    public apiService = new ApiService();
    public adminService =  new AdminService(this.apiService);
    public deviceService =  new DeviceService(this.apiService);
    public eventService = new EventService();
    public locationService =  new LocationService(this.apiService);
    public sensorService =  new SensorService(this.apiService);
}
