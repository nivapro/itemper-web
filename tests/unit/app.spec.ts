
import Vue from 'vue';

import App from '@/app.vue';
import {log} from '@/services/logger';

import Sensors from '@/components/sensors.vue';

import Vuetify from 'vuetify';

import { createLocalVue, mount } from '@vue/test-utils';

// import { shallow } from '@vue/test-utils';

describe('app.spec.ts', () => {
  const wrp: any = [];

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(Vuetify);

    // wrp = shallow(App, {
    //   localVue,
    // });
    // vm.$el === document.getElementById('app')

  });

  it('must have a app id', () => {
    // No items to display.
    // NOTE: By default, our component has some items, but we use
    // setData() here to override it with an empty array of items to
    // make sure we have not items at all and our test works as
    // expected.

    expect(wrp.attributes().id).toBe('app');
  });

  it('has the expected html structure', () => {
    expect(wrp.element).toMatchSnapshot();
  });
});
