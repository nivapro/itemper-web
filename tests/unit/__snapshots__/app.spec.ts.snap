// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app.spec.ts has the expected html structure 1`] = `
<div
  id="app"
>
  <!---->
</div>
`;
